include/revng/StackAnalysis/ABI/CallingConventionTrait.h

```diff
@@ -0,0 +1,1089 @@
+  /// pointers-to-copy, otherwise, they are placed on the stack.
+  ///
+  /// \note: this only affects the arguments with size exceeding the size of
+  /// a single stack "slot" (which is equal to the stack alignment).
+  static constexpr bool UsePointerToCopyForStackArguments = false;
+
+  /// Specifies whether vector registers are always used in rigid pairs, meaning
+  /// that an odd-index vector register cannot be used without the even-index
+  /// register from the same pair being used to hold half the value.
+  ///
+  /// \note: The option still is not fully implemented since the vector register
+  /// support is still super limited.
+  static constexpr bool VectorRegistersArePaired = false;
+
+  /// States whether specified ABI allows accepting aggregate object arguments
+  /// (structs, unions, etc) in the designated registers if they provide enough
+  /// space (including padding) and the maximum number of registers allowed for
+  /// each of the arguments.
+  ///
+  /// `0` means that using registers for ANY aggregate registers is disallowed,
+  /// even the object could fit into a single general purpose register.
+  ///
+  /// `1` means that passing values is only allowed if their size (including
+  /// padding) doesn't exceed the size of a single general purpose register.
+  ///
+  /// `2` means that passing aggregate values is only allowed if their size
+  /// (including padding) doesn't exceed the size of two general purpose
+  /// registers.
+  ///
+  /// ... and so on.
+  ///
+  /// if `MaximumGeneralPurposeRegistersPerArgument` is greater than or
+  /// equal to `GeneralPurposeArgumentRegisters.size()`, all the general purpose
+  /// registers allowed to be used for function arguments could be used for
+  /// a single aggregate argument.
+  ///
+  /// If the argument that doen't fit into the allowed register count it is
+  /// placed on the stack.
+  ///
+  /// \note: some special cases still apply. For example C++ abi doesn't allow
+  /// structs and classes that are not 'trivial for the purposes of calls' to
+  /// be passed in such a fashion. As such, object with irregular padding,
+  /// non-trivial construction/destruction, etc are only passed using the stack.
+  /// (see https://itanium-cxx-abi.github.io/cxx-abi)
+  static constexpr size_t MaxGeneralPurposeRegistersPerAggregateArgument = 0;
+
+  /// States whether specified ABI allows returning aggregate objects (structs
+  /// unions, etc) in the designated registers if they provide enough space
+  /// space (including padding) and the maximum number of registers allowed for
+  /// those values.
+  ///
+  /// The rules from `MaxGeneralPurposeRegistersPerAggregateArgument` apply,
+  /// see relevant comment for additional information.
+  /// \sa MaxGeneralPurposeRegistersPerAggregateArgument
+  static constexpr size_t MaxGeneralPurposeRegistersPerAggregateReturnValue = 0;
+
+  /// States whether specified ABI allows accepting primite object arguments
```

Comment:
Wait, as far as my understanding is concerned, `primitive` also includes pointers. Since they are, in their essence, just numbers. And we don't exactly care about pointed-to types at this point.

Reply:
In rev.ng mode parlance, *primitive* is not a pointer. A *scalar* is either a primitive or a pointer.
Let's use a unified lingo, ours. After all, the purpose of this file is to unify all the differences of the various ABIs into the model world.

Reply:
I see. Then a whole bunch of stuff would need to get renamed to scalars. I'll take a look on Monday.

Reply:
Updates?

Reply:
Done, forgot to close this thread, sorry.